File: test/oracle_simulator.js
// Copyright (c) 2021 Kentaro Hara
//
// This software is released under the MIT License.
// http://opensource.org/licenses/mit-license.php

const { deployProxy } = require('@openzeppelin/truffle-upgrades');
const JohnLawCoin = artifacts.require("JohnLawCoin");
const JohnLawBond = artifacts.require("JohnLawBond");
const Oracle = artifacts.require("Oracle");
const OracleForTesting = artifacts.require("OracleForTesting");
const common = require("./common.js");
const should_throw = common.should_throw;
const mod = common.mod;
const randint = common.randint;

contract("OracleSimulator", function (accounts) {
  let args = common.custom_arguments();
  assert.isTrue(args.length == 5);
  parameterized_test(accounts,
                     args[0],
                     args[1],
                     args[2],
                     args[3],
                     args[4]);
});

function parameterized_test(accounts,
                            _level_max,
                            _reclaim_threshold,
                            _proportional_reward_rate,
                            _voter_count,
                            _iteration) {
  let test_name = "Oracle parameters: " + "level_max=" + _level_max +
      " reclaim=" + _reclaim_threshold +
      " prop=" + _proportional_reward_rate +
      " voter_count=" + _voter_count +
      " iteration=" + _iteration;
  console.log(test_name);
  assert.isTrue(_voter_count <= accounts.length - 1);

  it(test_name, async function () {
    let _prev_tax = 0;
    let _coin = await deployProxy(JohnLawCoin, []);
    let _oracle = await deployProxy(OracleForTesting, []);
    await _oracle.overrideConstants(_level_max, _reclaim_threshold,
                                    _proportional_reward_rate);
    common.print_contract_size(_oracle, "OracleForTesting");

    let initial_coin_supply = (await _coin.totalSupply()).toNumber();

    for (let iter = 0; iter < _iteration; iter++) {
      console.log(iter);
      let voters = [];
      for (let i = 0; i < _voter_count; i++) {
        let voter = {
          address: accounts[i + 1],
          committed: false,
          deposit: 0,
          committed_level: 0,
          committed_salt: 0,
          committed_correctly: false,
          revealed: false,
          revealed_correctly: false,
          oracle_level: 0,
          salt: 0,
          reclaimed: false
        };
        voters.push(voter);
      }

      for (let i = 0; i < voters.length; i++) {
        assert.equal(voters[i].address, accounts[i + 1]);
        voters[i].committed = (randint(0, 99) < 95);
        if (voters[i].committed) {
          voters[i].deposit = randint(0, 10);
          voters[i].committed_level = randint(0, _level_max);
          voters[i].committed_salt = randint(0, 10);
          await _coin.mint(voters[i].address, voters[i].deposit);
          await check_commit(
              voters[i].address,
              await _oracle.encrypt(voters[i].address,
                                 voters[i].committed_level,
                                 voters[i].committed_salt),
              voters[i].deposit);

          assert.equal(await _coin.balanceOf(voters[i].address), 0);
          voters[i].committed_correctly = true;

          assert.equal(await _oracle.commit.call(
              _coin.address, voters[i].address,
              await _oracle.encrypt(voters[i].address,
                                 voters[i].committed_level,
                                 voters[i].committed_salt),
              0), false);
        }
      }

      let tax = randint(0, 200);
      await _coin.mint(await _coin.tax_account_(), tax);
      await check_advance(tax, _prev_tax);
      _prev_tax = tax;

      for (let i = 0; i < voters.length; i++) {
        assert.equal(voters[i].address, accounts[i + 1]);
        voters[i].revealed = (randint(0, 99) < 95);
        if (voters[i].revealed) {
          if (randint(0, 99) < 95) {
            voters[i].oracle_level = voters[i].committed_level;
          } else {
            voters[i].oracle_level = randint(0, _level_max);
          }
          if (randint(0, 99) < 95) {
            voters[i].salt = voters[i].committed_salt;
          } else {
            voters[i].salt = randint(0, 10);
          }
          voters[i].revealed_correctly = (
              voters[i].committed_correctly &&
                voters[i].oracle_level == voters[i].committed_level &&
                0 <= voters[i].oracle_level &&
                voters[i].oracle_level < _level_max &&
                voters[i].salt == voters[i].committed_salt);
          if (voters[i].revealed_correctly) {
            await check_reveal(voters[i].address,
                               voters[i].oracle_level,
                               voters[i].salt);
          }
          assert.equal(await _oracle.reveal.call(
              voters[i].address, voters[i].oracle_level,
              voters[i].salt), false);
        }
      }

      let deposits = [];
      let counts = [];
      for (let level = 0; level < _level_max; level++) {
        deposits.push(0);
        counts.push(0);
      }
      let deposit_total = 0;
      for (let i = 0; i < voters.length; i++) {
        if (voters[i].committed_correctly) {
          deposit_total += voters[i].deposit;
        }
        if (voters[i].revealed_correctly) {
          deposits[voters[i].oracle_level] += voters[i].deposit;
          counts[voters[i].oracle_level] += 1;
        }
      }

      let max_deposit = 0;
      let max_count = 0;
      let mode_level = _level_max;
      for (let level = 0; level < _level_max; level++) {
        if (counts[level] > 0 &&
            (mode_level == _level_max ||
             max_deposit < deposits[level] ||
             (max_deposit == deposits[level] &&
              max_count < counts[level]))) {
          max_deposit = deposits[level];
          max_count = counts[level];
          mode_level = level;
        }
      }

      tax = randint(0, 200);
      let deposit_to_reclaim = 0;
      if (mode_level == _level_max) {
        reward_total = deposit_total + tax;
      } else {
        for (let level = 0; level < _level_max; level++) {
          if (mode_level - _reclaim_threshold <= level &&
              level <= mode_level + _reclaim_threshold) {
            deposit_to_reclaim += deposits[level];
          }
        }
        reward_total = deposit_total - deposit_to_reclaim + tax;
      }
      assert.equal(deposit_to_reclaim + reward_total,
                   deposit_total + tax);

      await _coin.mint(await _coin.tax_account_(), tax);
      await check_advance(tax, _prev_tax);
      assert.equal(await _oracle.getModeLevel(), mode_level);
      _prev_tax = tax;

      let reclaim_total = 0;
      for (let i = 0; i < voters.length; i++) {
        assert.equal(voters[i].address, accounts[i + 1]);
        voters[i].reclaimed = (randint(0, 99) < 95);
        if (voters[i].reclaimed) {
          assert.equal(await _coin.balanceOf(voters[i].address), 0);
          let reclaimed = 0;
          let reward = 0;
          let should_reclaim = false;
          if ((voters[i].revealed_correctly &&
               voters[i].oracle_level == mode_level)) {
            assert.notEqual(mode_level, _level_max);
            if (deposits[mode_level] > 0) {
              reward += Math.trunc(
                  (_proportional_reward_rate * reward_total *
                   voters[i].deposit) / (100 * deposits[mode_level]));
            }
            reward += Math.trunc(
                ((100 - _proportional_reward_rate) * reward_total) /
                  (100 * counts[mode_level]));
            reclaimed = voters[i].deposit
            should_reclaim = true;
          } else if (voters[i].revealed_correctly &&
                     mode_level - _reclaim_threshold <=
                     voters[i].oracle_level &&
                     voters[i].oracle_level <=
                     mode_level + _reclaim_threshold) {
            assert.notEqual(mode_level, _level_max);
            reclaimed = voters[i].deposit;
            should_reclaim = true;
          }
          if (should_reclaim) {
            await check_reclaim(voters[i].address, reclaimed, reward);
          }
          common.array_equal(await _oracle.reclaim.call(
              _coin.address, voters[i].address), [0, 0]);
          reclaim_total += reclaimed + reward;
          assert.equal(await _coin.balanceOf(voters[i].address),
                       reclaimed + reward);
          await _coin.burn(voters[i].address, reclaimed + reward);
        }
      }

      assert.equal(deposit_to_reclaim + reward_total, deposit_total + tax);
      let remainder = deposit_total + tax - reclaim_total;
      tax = randint(0, 200);
      await _coin.mint(await _coin.tax_account_(), tax);
      await check_advance(tax, remainder);
      _prev_tax = tax;

      assert.equal(await _coin.totalSupply(), _prev_tax + initial_coin_supply);
    }

    async function check_commit(account, hash, deposit) {
      await _coin.transferOwnership(_oracle.address);
      let receipt =
          await _oracle.commit(_coin.address, account, hash, deposit);
      await _oracle.revokeOwnership(_coin.address);
      let args = receipt.logs.filter(e => e.event == 'CommitEvent')[0].args;
      assert.equal(args.sender, account);
      assert.equal(args.hash, hash);
      assert.equal(args.deposited, deposit);
    }

    async function check_reveal(account, level, salt) {
      let receipt = await _oracle.reveal(account, level, salt);
      let args = receipt.logs.filter(e => e.event == 'RevealEvent')[0].args;
      assert.equal(args.sender, account);
      assert.equal(args.oracle_level, level);
      assert.equal(args.salt, salt);
    }

    async function check_reclaim(account, reclaimed, reward) {
      await _coin.transferOwnership(_oracle.address);
      let receipt = await _oracle.reclaim(_coin.address, account);
      await _oracle.revokeOwnership(_coin.address);
      let args = receipt.logs.filter(e => e.event == 'ReclaimEvent')[0].args;
      assert.equal(args.sender, account);
      assert.equal(args.reclaimed, reclaimed);
      assert.equal(args.rewarded, reward);
    }

    async function check_advance(tax, burned) {
      await _coin.transferOwnership(_oracle.address);
      let receipt = await _oracle.advance(_coin.address);
      await _oracle.revokeOwnership(_coin.address);
      let args = receipt.logs.filter(
          e => e.event == 'AdvancePhaseEvent')[0].args;
      assert.isTrue(args.epoch_id >= 3);
      assert.equal(args.tax, tax);
      assert.equal(args.burned, burned);
    }

  });

}

File: python/oracle_simulator.py
#!/usr/bin/env python3
#
# Copyright (c) 2021 Kentaro Hara
#
# This software is released under the MIT License.
# http://opensource.org/licenses/mit-license.php

from johnlawcoin import *
import unittest, random

class OracleSimulator(unittest.TestCase):
    def __init__(self, level_max, reclaim_threshold,
                 proportional_reward_rate, voter_count, iteration):
        super().__init__()
        print("level_max=%d reclaim=%d prop=%d voter_count=%d iter=%d" %
              (level_max, reclaim_threshold,
               proportional_reward_rate, voter_count, iteration))
        self._level_max = level_max
        self._reclaim_threshold = reclaim_threshold
        self._proportional_reward_rate = proportional_reward_rate
        self._voter_count = voter_count
        self._iteration = iteration

        self._coin = JohnLawCoin(0)
        self._oracle = Oracle()
        self._oracle.override_constants_for_testing(
            self._level_max, self._reclaim_threshold,
            self._proportional_reward_rate)

        self._prev_tax = 0

    def teardown(self):
        pass

    def run(self):
        initial_coin_supply = self._coin.total_supply
        for i in range(self._iteration):
            self.run_cycle()
            self.assertEqual(self._coin.total_supply,
                             self._prev_tax + initial_coin_supply)

    def run_cycle(self):

        class Voter:
            def __init__(self, address):
                self.address = address
                self.committed = False
                self.deposit = 0
                self.committed_level = 0
                self.committed_salt = 0
                self.committed_correctly = False
                self.revealed = False
                self.revealed_correctly = False
                self._oracle_level = 0
                self.salt = 0
                self.reclaimed = False

        voters = []
        for i in range(self._voter_count):
            voters.append(Voter(i + 1))

        for i in range(len(voters)):
            self.assertEqual(voters[i].address, i + 1)
            voters[i].committed = (random.randint(0, 99) < 95)
            if voters[i].committed:
                voters[i].deposit = random.randint(0, 10)
                voters[i].committed_level = random.randint(0, self._level_max)
                voters[i].committed_salt = random.randint(0, 10)
                self._coin.mint(voters[i].address, voters[i].deposit)
                result = self._oracle.commit(
                    self._coin, voters[i].address,
                    Oracle.encrypt(voters[i].address,
                                   voters[i].committed_level,
                                   voters[i].committed_salt),
                    voters[i].deposit)

                self.assertEqual(result, True)
                self.assertEqual(self._coin.balance_of(voters[i].address), 0)
                voters[i].committed_correctly = True

                self.assertEqual(self._oracle.commit(
                    self._coin, voters[i].address,
                    Oracle.encrypt(voters[i].address,
                                   voters[i].committed_level,
                                   voters[i].committed_salt), 0), False)

        tax = random.randint(0, 200)
        self._coin.mint(self._coin.tax_account, tax)
        burned = self._oracle.advance(self._coin)
        self.assertEqual(burned, self._prev_tax)
        self._prev_tax = tax

        for i in range(len(voters)):
            self.assertEqual(voters[i].address, i + 1)
            voters[i].revealed = (random.randint(0, 99) < 95)
            if voters[i].revealed:
                if random.randint(0, 99) < 95:
                    voters[i].oracle_level = voters[i].committed_level
                else:
                    voters[i].oracle_level = random.randint(
                        0, self._level_max)
                if random.randint(0, 99) < 95:
                    voters[i].salt = voters[i].committed_salt
                else:
                    voters[i].salt = random.randint(0, 10)
                voters[i].revealed_correctly = (
                    voters[i].committed_correctly and
                    voters[i].oracle_level == voters[i].committed_level and
                    0 <= voters[i].oracle_level and
                    voters[i].oracle_level < self._level_max and
                    voters[i].salt == voters[i].committed_salt)
                result = self._oracle.reveal(
                    voters[i].address,
                    voters[i].oracle_level,
                    voters[i].salt)
                self.assertEqual(result, voters[i].revealed_correctly)
                self.assertEqual(
                    self._oracle.reveal(voters[i].address,
                                       voters[i].oracle_level,
                                       voters[i].salt), False)
                self.assertEqual(
                    self._oracle.reveal(-voters[i].address,
                                       voters[i].oracle_level,
                                       voters[i].salt), False)

        deposits = [0] * self._level_max
        counts = [0] * self._level_max
        deposit_total = 0
        for i in range(len(voters)):
            if voters[i].committed_correctly:
                deposit_total += voters[i].deposit
                assert(voters[i].deposit >= 0)
            if voters[i].revealed_correctly:
                deposits[voters[i].oracle_level] += voters[i].deposit
                counts[voters[i].oracle_level] += 1

        max_deposit = 0
        max_count = 0
        mode_level = self._level_max
        for level in range(self._level_max):
            if (counts[level] > 0 and
                (mode_level == self._level_max or
                 max_deposit < deposits[level] or
                 (max_deposit == deposits[level] and
                  max_count < counts[level]))):
                max_deposit = deposits[level]
                max_count = counts[level]
                mode_level = level

        tax = random.randint(0, 200)
        deposit_to_reclaim = 0
        if mode_level == self._level_max:
            reward_total = deposit_total + tax
        else:
            for level in range(self._level_max):
                if (mode_level - self._reclaim_threshold <= level and
                    level <= mode_level + self._reclaim_threshold):
                    deposit_to_reclaim += deposits[level]
            reward_total = deposit_total - deposit_to_reclaim + tax
        self.assertEqual(deposit_to_reclaim + reward_total,
                         deposit_total + tax)

        self._coin.mint(self._coin.tax_account, tax)
        burned = self._oracle.advance(self._coin)
        self.assertEqual(self._oracle.get_mode_level(), mode_level)
        self.assertEqual(burned, self._prev_tax)
        self._prev_tax = tax

        reclaim_total = 0
        for i in range(len(voters)):
            self.assertEqual(voters[i].address, i + 1)
            voters[i].reclaimed = (random.randint(0, 99) < 95)
            if voters[i].reclaimed:
                self.assertEqual(self._coin.balance_of(voters[i].address), 0)
                reward = 0
                reclaimed = 0
                if (voters[i].revealed_correctly and
                    voters[i].oracle_level == mode_level):
                    self.assertNotEqual(mode_level, self._level_max)
                    if deposits[mode_level] > 0:
                        reward += int(
                            (self._proportional_reward_rate * reward_total *
                             voters[i].deposit) /
                            (100 * deposits[mode_level]))
                    reward += int(
                        ((100 - self._proportional_reward_rate)
                         * reward_total) /
                        (100 * counts[mode_level]))
                    reclaimed = voters[i].deposit
                elif (voters[i].revealed_correctly and
                      mode_level - self._reclaim_threshold <=
                      voters[i].oracle_level and
                      voters[i].oracle_level <=
                      mode_level + self._reclaim_threshold):
                    self.assertNotEqual(mode_level, self._level_max)
                    reclaimed = voters[i].deposit
                self.assertEqual(self._oracle.reclaim(
                    self._coin, voters[i].address), (reclaimed, reward))
                reclaim_total += reclaimed + reward
                self.assertEqual(self._coin.balance_of(voters[i].address),
                                 reclaimed + reward)
                self._coin.burn(voters[i].address, reclaimed + reward)

                self.assertEqual(
                    self._oracle.reclaim(self._coin, voters[i].address), (0, 0))
                self.assertEqual(self._oracle.reclaim(
                    self._coin, -voters[i].address), (0, 0))

        self.assertEqual(deposit_to_reclaim + reward_total,
                         deposit_total + tax)
        burned = deposit_total + tax - reclaim_total
        tax = random.randint(0, 200)
        self._coin.mint(self._coin.tax_account, tax)
        self.assertEqual(self._oracle.advance(self._coin), burned)
        self._prev_tax = tax


def main():
    iteration = 1000
    for level_max in [2, 3, 4, 9]:
        for reclaim_threshold in range(0, level_max):
            for proportional_reward_rate in [0, 1, 90, 100]:
                for voter_count in [0, 1, 100]:
                    test = OracleSimulator(level_max,
                                           reclaim_threshold,
                                           proportional_reward_rate,
                                           voter_count,
                                           iteration)
                    test.run()
                    test.teardown()


if __name__ == "__main__":
    main()
